src/fsharp/utils/FileSystem.fs

```diff
@@ -1,164 +1,778 @@
+                        try
+                            use stream = mmf.CreateViewStream(0L, length, MemoryMappedFileAccess.ReadWrite)
+                            bytes.CopyTo stream
+                            Some mmf
+                        with
+                        | _ ->
+                            mmf.Dispose()
+                            None
+                    with
+                    | _ ->
+                        None
+
+[<RequireQualifiedAccess>]
+module internal FileSystemUtils =
+    let checkPathForIllegalChars  =
+        let chars = new System.Collections.Generic.HashSet<_>(Path.GetInvalidPathChars())
+        (fun (path:string) ->
+            for c in path do
+                if chars.Contains c then raise(IllegalFileNameChar(path, c)))
+
+    let checkSuffix (x:string) (y:string) = x.EndsWithOrdinal(y)
+
+    let hasExtensionWithValidate (validate:bool) (s:string) =
+        if validate then (checkPathForIllegalChars s) |> ignore
+        let sLen = s.Length
+        (sLen >= 1 && s.[sLen - 1] = '.' && s <> ".." && s <> ".")
+        || Path.HasExtension(s)
+
+    let hasExtension (s:string) = hasExtensionWithValidate true s
+
+    let chopExtension (s:string) =
+        checkPathForIllegalChars s
+        if s = "." then "" else // for OCaml compatibility
+        if not (hasExtensionWithValidate false s) then
+            raise (System.ArgumentException("chopExtension")) // message has to be precisely this, for OCaml compatibility, and no argument name can be set
+        Path.Combine (Path.GetDirectoryName s, Path.GetFileNameWithoutExtension(s))
+
+    let fileNameOfPath s =
+        checkPathForIllegalChars s
+        Path.GetFileName(s)
+
+    let fileNameWithoutExtensionWithValidate (validate:bool) s =
+        if validate then checkPathForIllegalChars s |> ignore
+        Path.GetFileNameWithoutExtension(s)
+
+    let fileNameWithoutExtension s = fileNameWithoutExtensionWithValidate true s
+
+    let trimQuotes (s:string) =
+        s.Trim( [|' '; '\"'|] )
+
+    let hasSuffixCaseInsensitive suffix filename = (* case-insensitive *)
+        checkSuffix (String.lowercase filename) (String.lowercase suffix)
+
+    let isDll file = hasSuffixCaseInsensitive ".dll" file
+
+type IAssemblyLoader =
+    abstract AssemblyLoadFrom: fileName: string -> Assembly
+    abstract AssemblyLoad: assemblyName: AssemblyName -> Assembly
+
+type DefaultAssemblyLoader() =
+    interface IAssemblyLoader with
+        member _.AssemblyLoadFrom(fileName: string) = Assembly.UnsafeLoadFrom fileName
+        member _.AssemblyLoad(assemblyName: AssemblyName) = Assembly.Load assemblyName
+
+type IFileSystem =
+    abstract AssemblyLoader: IAssemblyLoader
+
+    abstract OpenFileForReadShim: filePath: string * ?useMemoryMappedFile: bool * ?shouldShadowCopy: bool -> ByteMemory
```

Comment:
Not sure `useMemoryMappedFile` needs to be in the signature since it is the implementation details.
But we somehow need to indicate whether we want to use mmap or not (we only want to use it for for big files, it doesn't make sense to use it for source files since it locks the file and we need to either reuse the map when accessing from another process or shadow copy all source files).

Reply:
We discussed this a bit today. I think it's fine that you can't open a file if it's memory-mapped and not shadow copied; we expect that.

We will iterate more on this when we need to, but otherwise it looks good.